capabilityreceiver/cmd/migratevalidations/main.go

```diff
@@ -0,0 +1,498 @@
+			URL: ha.fhirURL,
+		}
+		ha.result <- result
+		return nil
+	}
+	defer historyRows.Close()
+	var validationVals []validationArgs
+	for historyRows.Next() {
+		var val validationArgs
+		err = historyRows.Scan(&val.capStatByte,
+			&val.tlsVersion,
+			pq.Array(&val.mimeTypes),
+			&val.httpResponse,
+			&val.smartHttpResponse,
+			&val.updatedTime)
+		if err != nil {
+			log.Warnf("Error while scanning the rows of the history table for URL %s. Error: %s", ha.fhirURL, err)
+			continue
+		}
+		validationVals = append(validationVals, val)
+	}
+
+	for _, val := range validationVals {
+		// Create the capability statement object
+		var capStat capabilityparser.CapabilityStatement
+		var capInt map[string]interface{}
+		if len(val.capStatByte) > 0 {
+			err = json.Unmarshal(val.capStatByte, &capInt)
+			if err != nil {
+				log.Warnf("Error while unmarshalling the rows of the history table for URL %s. Error: %s", ha.fhirURL, err)
+			}
+
+			capStat, err = capabilityparser.NewCapabilityStatementFromInterface(capInt)
+			if err != nil {
+				log.Warnf("unable to parse CapabilityStatement out of message with url %s. Error: %s", ha.fhirURL, err)
+			}
+		}
+
+		// @TODO update this based on Emily's PR
+		fhirVersion := ""
+		if capStat != nil {
+			fhirVersion, _ = capStat.GetFHIRVersion()
+		}
+
+		validator := validation.ValidatorForFHIRVersion(fhirVersion)
+		validationObj := validator.RunValidation(capStat, val.httpResponse, val.mimeTypes, fhirVersion, val.tlsVersion, val.smartHttpResponse)
+		// Create ID in validation_results table and then get the ID
+		var valResID int
+		valRow, err := addValidationResultStatement.QueryContext(ctx)
+		if err == nil {
+			defer valRow.Close()
+			count := 0
+			for valRow.Next() {
+				if count > 0 {
+					log.Warnf("More than 1 ID added for URL %s", ha.fhirURL)
+					break
+				}
+				err = valRow.Scan(&valResID)
+				count++
+			}
+		}
+		if err != nil {
+			log.Warnf("Failed to add a new ID. Error: %s", err)
+			result := Result{
+				URL: ha.fhirURL,
+			}
+			ha.result <- result
+			return nil
+		}
+		// Then add each element of the validationObj array as a row to the table with that id
+		for _, ruleInfo := range validationObj.Results {
+			_, err = addValidationStatement.ExecContext(ctx,
+				ruleInfo.RuleName,
+				ruleInfo.Valid,
+				ruleInfo.Expected,
+				ruleInfo.Actual,
+				ruleInfo.Comment,
+				ruleInfo.Reference,
+				ruleInfo.ImplGuide,
+				valResID)
+			if err != nil {
+				log.Warnf("Failed to add validation for URL %s. Error: %s", ha.fhirURL, err)
+				result := Result{
+					URL: ha.fhirURL,
+				}
+				ha.result <- result
+				return nil
+			}
+		}
+		// Then have to update the row in the history table with that id
+		_, err = updateFHIREndpointValResStatement.ExecContext(ctx, valResID, val.updatedTime, ha.fhirURL)
+		if err != nil {
+			log.Warnf("Error while updating the row of the table for URL %s at %s. Error: %s", ha.fhirURL, val.updatedTime.String(), err)
+			result := Result{
```

Comment:
I've seen the add result to ha.result code a few times -- might be worth adding an addResult() function to "DRY" out the code.

Reply:
done!